rsyslog/8.2106.0/rsyslog-8.2106.0.patch

```diff
@@ -0,0 +1,455 @@
+@@ -417,10 +438,17 @@ osslGlblInit(void)
+ 	const char *caFile, *certFile, *keyFile;
+ 
+ 	/* Setup OpenSSL library */
++	#ifndef ENABLE_WOLFSSL
+ 	if((opensslh_THREAD_setup() == 0) || !SSL_library_init()) {
++	#else
++	if(!SSL_library_init()) {
++	#endif
+ 		LogError(0, RS_RET_NO_ERRCODE, "Error: OpenSSL initialization failed!");
+ 	}
+ 
++	#if defined(ENABLE_WOLFSSL) && defined(DEBUG_WOLFSSL)
++	wolfSSL_Debugging_ON();
++	#endif
+ 	/* Load readable error strings */
+ 	SSL_load_error_strings();
+ 	ERR_load_BIO_strings();
+@@ -452,8 +480,16 @@ osslGlblInit(void)
+ 		bHaveKey = 1;
+ 	}
+ 
++/* wolfSSL will use TLS 1.3 if it's compiled in and we call SSLv23_method. This
++ * is at odds with the fact that rsyslog allows usage of anonymous cipher
++ * suites, which were deprecated in TLS 1.3. To continue to allow these suites,
++ * we explicitly request TLS 1.2 here. */
++#ifdef ENABLE_WOLFSSL
++	ctx = SSL_CTX_new(TLSv1_2_method());
++#else
+ 	/* Create main CTX Object */
+ 	ctx = SSL_CTX_new(SSLv23_method());
++#endif
+ 	if(bHaveCA == 1 && SSL_CTX_load_verify_locations(ctx, caFile, NULL) != 1) {
+ 		LogError(0, RS_RET_TLS_CERT_ERR, "Error: CA certificate could not be accessed. "
+ 				"Check at least: 1) file path is correct, 2) file exist, "
+@@ -501,6 +537,11 @@ finalize_it:
+ static rsRetVal
+ osslAnonInit(void)
+ {
++	#ifdef ENABLE_WOLFSSL
++	DH* dh;
++	BIGNUM* p;
++	BIGNUM* g;
++	#endif
+ 	DEFiRet;
+ 	pthread_mutex_lock(&anonInit_mut);
+ 	if (bAnonInit == 1) {
+@@ -509,7 +550,19 @@ osslAnonInit(void)
+ 	}
+ 	dbgprintf("osslAnonInit Init Anon OpenSSL helpers\n");
+ 
+-	#if OPENSSL_VERSION_NUMBER >= 0x10002000L
++	#ifdef ENABLE_WOLFSSL
++	/* wolfSSL doesn't have support for ECDH anonymous ciphers, just DH.
++	 * Accordingly, we need to generate an ephemeral DH key. This is done using
++	 * the parameters from RFC3526 prime 2048 (id 14).*/
++	dh = DH_new();
++	p = get_rfc3526_prime_2048(NULL);
++	g = BN_new();
++	BN_set_word(g, 2);
++	DH_set0_pqg(dh, p, NULL, g);
++	DH_generate_key(dh);
++	SSL_CTX_set_tmp_dh(ctx, dh);
++	DH_free(dh);
++	#elif OPENSSL_VERSION_NUMBER >= 0x10002000L
+ 	/* Enable Support for automatic EC temporary key parameter selection. */
+ 	SSL_CTX_set_ecdh_auto(ctx, 1);
+ 	#else
+@@ -636,8 +689,21 @@ osslInitSession(nsd_ossl_t *pThis) /* , nsd_ossl_t *pServer) */
+ 	}
+ 
+ 	if (bAnonInit == 1) { /* no mutex needed, read-only after init */
++		#ifdef ENABLE_WOLFSSL
++		if (bHaveCert == 0) {
++			/* To get wolfSSL to use the anonymous ciphers, they need to be
++			 * moved to the front of the list (or be the only items in the
++			 * list). But, we only want to do this if we don't have a cert.
++			 * Otherwise, we want to prefer non-anonymous ciphers. This matches
++			 * the OpenSSL behavior implied below. */
++			strncpy(pristringBuf, "ADH-AES256-GCM-SHA384:ADH-AES128-SHA",
++				sizeof(pristringBuf));
++		}
++		else {
++			strncpy(pristringBuf, "ALL", sizeof(pristringBuf));
++		}
```

Comment:
Why do you think that OpenSSL implies that it prefers anon ciphers? The anon ciphers (`ADH`)  are added after `ALL` and `COMPLEMENTOFDEFAULT`. IIRC this would be the order of preference.

Reply:
That's correct. However, OpenSSL is smart enough to recognize that it doesn't have a certificate and only pick from the available anonymous ciphers as a result. wolfSSL isn't able to do this. It will go for whatever is at the front of the preference list, regardless of whether or not a cert has been loaded. Then, when the server tries to send the cert in the handshake, it fails because there's no cert to send. `if (bHaveCert == 0)` --> the only ciphers we could reasonably pick are anonymous, so just set the list to the anonymous ones.

I initially had changes to make this work in wolfSSL (i.e. detect no cert, limit cipher list accordingly), but Kaleb and I discussed on Slack and decided that wasn't the optimal solution. I can link you to the thread if you're interested.

Reply:
Thanks for the explanation. Add a short comment explaining that the preference of anon ciphers come from the lack of cert.

Reply:
Sure thing. I feel like that is conveyed by the comment at line 246 already, but feel free to propose a rewording if you think it's unclear and I'll change it.

Reply:
The implied below part messed me up. Let's substitute that for something like "OpenSSL defaults to anon ciphers when a cert is not available".